store/tikv/region_cache.go

```diff
@@ -246,7 +241,43 @@ func (c *RegionCache) loadRegion(bo *Backoffer, key []byte) (*Region, error) {
 	}
 }
 
-// llrbItem is llrbTree's Item that uses []byte for compare.
+// OnRegionStale removes the old region and inserts new regions into the cache.
+func (c *RegionCache) OnRegionStale(old *Region, newRegions []*metapb.Region) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.dropRegionFromCache(old.VerID())
+
+	for _, meta := range newRegions {
+		if err := decodeRegionMetaKey(meta); err != nil {
+			log.Errorf("newRegion's range key is not encoded: %v, %v", meta, err)
+			continue
+		}
+		reorderRegionPeers(meta, old.peer.GetStoreId())
+		peer := meta.Peers[0]
+		if peer.GetStoreId() != old.peer.GetStoreId() {
+			continue
+		}
+		c.insertRegionToCache(&Region{
+			meta: meta,
+			peer: peer,
+			addr: old.GetAddress(),
+		})
+	}
+}
+
+// reorderRegionPeers moves the leader peer to the first and makes it easier to
+// try the next peer if the current peer does not respond.
+func reorderRegionPeers(r *metapb.Region, leaderStoreID uint64) {
```

Comment:
Will the leaderStoreID always found in peers?
If not, should we returns a bool indicates if leaderStoreID has been moved to first?